### ISSUE TITLE
Research: counterexample-guided synthesis of bounded Agent Interface skills

### ISSUE THREAD

**Issue by Unjuno** [Unjuno/agent-interface#4262]
## Motivation

Related: #857 macro compilation, #696 sensorimotor graphs, #751 skill registry, #4158 evidence-bounded delegation, #3033 experiment discipline.

The repository already separates candidate construction from independent validation. A natural next step is to make this loop explicit at the **skill** level:

```text
Astora/Astra proposes bounded skill
        ->
independent verifier finds counterexample
        ->
candidate is repaired
        ->
revalidate on accumulated counterexamples
```

This resembles counterexample-guided inductive synthesis (CEGIS).

The goal is not unrestricted self-modifying agents. It is to test whether a small typed skill can be improved through explicit retained counterexamples while preserving negative evidence.

Current-main intake: `affe58482c57317f135b15b65f8e94beb8fe42ac`.

Targeted duplicate search found no CEGIS/counterexample-guided skill-synthesis Issue.

## H

For one bounded semantic operation, a candidate macro/graph synthesized from positive examples can be iteratively repaired using independently generated counterexamples, reaching the frozen skill contract with fewer manual rule additions than a one-shot candidate, while preserving every prior failing counterexample.

## T

No live user desktop. Controlled fixture/generator.

Freeze:
- one small skill grammar (guards, branches, bounded actions, YIELD);
- one initial positive demonstration set;
- independent verifier/counterexample generator;
- maximum synthesis rounds;
- exact acceptance contract;
- prohibition on changing the task/oracle.

Arms:
1. `ONE_SHOT_SKILL`;
2. `COUNTEREXAMPLE_GUIDED_SKILL`.

Each repair round may modify only the candidate skill. The verifier/oracle and previous counterexamples are immutable.

Required counterexample classes:
- wrong target identity;
- missing evidence;
- alternate valid state requiring another branch;
- forbidden effect;
- stale/generation mismatch;
- case requiring YIELD.

## Measurements

- rounds to contract satisfaction;
- retained counterexamples passed;
- new regressions;
- skill size/branch count;
- verifier calls;
- frontier/Astora calls if used;
- held-out correctness;
- YIELD correctness;
- synthesis wall time.

## D

- `PASS_CEGIS_SKILL_SYNTHESIS_SCOPED` — candidate reaches all frozen gates within the round/size budget, preserves all accumulated counterexamples, and improves held-out skill correctness over one-shot construction.
- `FAIL_COUNTEREXAMPLE_REGRESSION`.
- `FAIL_SYNTHESIS_BRANCH_EXPLOSION`.
- `HOLD_ONE_SHOT_ALREADY_SUFFICIENT`.
- `HOLD_VERIFIER_NOT_INDEPENDENT`.
- `STOP_ORACLE_OR_GRAMMAR_NOT_FROZEN`.

## C

- The verifier may merely teach the exact test set.
- Repair can overfit counterexamples.
- A deterministic human-authored macro may be simpler.
- Skill grammar choice may determine success more than the synthesis loop.

## U

No hidden chain-of-thought required. Only externally observable candidate artifacts, counterexamples, actions/effects, and validation results enter the loop. No live promotion follows from a synthesis PASS.

**Comment by Unjuno** [comment]
CLAIM / RESERVATION — 2026-09-23.

Base main: `62bb46cf55761fce72b06d0df8bbacd620a501c1`.
Owned branch: `research/cegis-skill-4262-20260923`.
Owned additive path: `research/analysis/cegis_skill_4262_v1/**`.

First bounded rung:
- H: with one fixed typed skill grammar and immutable oracle, counterexample-guided resynthesis can reach the frozen contract within a small round/branch budget and improve held-out correctness over the same one-shot synthesizer trained only on initial positive demonstrations, while preserving every accumulated counterexample.
- T: stdlib-only finite six-Boolean state universe; outputs ACT_PRIMARY / ACT_ALT / YIELD. Fixed rule templates cover forbidden effect, stale generation, ambiguity, missing evidence, primary-target and alternate-target branches. Initial demos, verifier split, held-out split, candidate ordering/tie-break, max rounds and max rule count are frozen before formal. ONE_SHOT uses the same synthesizer once; CEGIS adds only the verifier's first deterministic counterexample each round.
- D: PASS only if CEGIS reaches exact verifier contract within <=6 counterexample rounds and <=6 rules, passes every accumulated counterexample, has zero forbidden/stale/missing-evidence unsafe actions, exact YIELD contract, and strictly higher held-out accuracy than ONE_SHOT. Independent raw auditor + >=10 corruption controls required. Regression/branch explosion/one-shot-sufficient classes remain as Issue-defined alternatives.
- C: typed grammar already contains the needed semantic guard concepts; this tests counterexample-guided search, not automatic discovery of new predicates.
- U: no model/provider/GUI/input/network/latency/token/product claim; synthesized skill has no execution authority.

Construction/debug excluded. Public exact source/hash/gate freeze precedes one formal enumeration; formal reruns/replacements/post-result tuning=0.

**Comment by Unjuno** [comment]
## Public preformal freeze — formal invocations 0

Allocation `cegis-skill-4262-20260923-01` frozen at commit `23909c8a3d1e8b3c43d959cd1465fe786690faf1`.

Exact Git blobs:
- PLAN `5a24a81f37eb15d58fa8c76ec47a83f209d4f83d` — SHA256 `db93a273035995c395fca9704e595f99d1b8e7f0b8848c01435536526da7b2c9`;
- ENV `005131f7a5c8c046b7594edd81f8bade2f56b34b`;
- CONSTRUCTION `782d60a1d540998dd66ac025a040e11bfc703314`;
- SOURCE_MANIFEST `af002893cab9d63997e634e952c71a4a9a59cfb9`;
- FREEZE `6ccce91806ac43b7a2af320c965bdd7660e21c4b`, local SHA256 `e0cd4692ad90e33e12e7bf219a5553cb95db1ae1c037f01a91da0e24bad39901`;
- SOURCE.tar.xz `52358bef747f8220a76b3af02ab2d17890cb6b43`, 4,728 bytes, SHA256 `55a1da35e063252bfb0c3c7c47dbeb58b5bcf7661e5844e775178e6eb33b5637`.

Excluded construction runner: ONE_SHOT held-out 5/16; CEGIS 16/16 with counterexamples `[1,17,7,30,51]`, six final rules, CEGIS unsafe-action-on-YIELD=0. Initial auditor passed unmodified raw, but the first corruption-control harness hit `KeyError` on a deliberately dropped row. That preformal auditor defect is retained; only the auditor was repaired to classify missing rows as an error. Scientific runner/oracle/synthesizer were unchanged. Repaired audit errors=[]; 12/12 mutations rejected.

Formal command fixed as `python -B runner.py --mode formal --out formal-01`, then frozen read-only audit/controls. Formal invocations/reruns remain **0/0** at this comment.

**Comment by Unjuno** [comment]
## First formal outcome — complete, no reruns

**PASS_CEGIS_SKILL_SYNTHESIS_SCOPED** for `cegis-skill-4262-20260923-01`.

Formal command executed exactly once after public source freeze; reruns/replacements/post-result tuning = 0/0/0.

- ONE_SHOT held-out: **5/16**;
- CEGIS held-out: **16/16**;
- deterministic verifier counterexamples: `[1,17,7,30,51]` (5);
- final CEGIS rule count: **6**;
- action on oracle-YIELD states across all64 states: ONE_SHOT **45**, CEGIS **0**;
- independent raw audit: 128 rows, `errors=[]`;
- 12/12 coherent evidence mutations rejected;
- every frozen scientific source hash matched `FREEZE.json` after formal;
- authority remained `none`.

Formal identities:
- RAW SHA256 `cc6355b673e71d166aeb8d623545a941ee312c03cb614b9a9af267a7ac226ba1`;
- formal RESULT SHA256 `d653a526ea4f3c936950797fe9cac9a172a1ffa342346480fecf55d6e55d1259`;
- AUDIT SHA256 `675b641681c0c5bb97f5cfa6682f62e7e9c1772ebd145a14f18ac5462c74b2ca`;
- CONTROLS SHA256 `e489ad005a0ddccd9ff64d406d1a41dd5f31505f6b5eeed9b5a51bdb9f8e9287`.

The excluded construction's original auditor passed unmodified raw but the first corruption-control harness raised `KeyError` on a deliberately dropped row. That preformal audit-harness STOP is retained exactly; only the auditor was repaired before freeze to classify missing rows as an error. Scientific runner/oracle/grammar/synthesizer were unchanged.

Lossless postformal evidence archive (formal + repaired construction + retained failed auditor history): 5,404 bytes, SHA256 `e64829c7080f8ae8d1dfe652c7758a2cc0c38db19ffbe402ab88d45cd74f41da`, 17 members.

Scope: this shows counterexample-guided search can find a safe rule ordering inside one finite grammar whose predicates already contain all required concepts. It does not demonstrate discovery of missing predicates, GUI/model skill generation, live authority, latency/token benefit or product readiness.
